internal/debug/dot.go

```diff
@@ -24,21 +25,33 @@ import (
 // quick and dirty dot debugging package
 
 type dotWriter struct {
-	*dag.Builder
+	Builder dag.BuilderType
 }
 
 type pair struct {
 	a, b interface{}
 }
 
+type nodeCollection map[interface{}]bool
+type edgeCollection map[pair]bool
+
 func (dw *dotWriter) writeDot(w io.Writer) {
+	nodes, edges := collectDag(dw.Builder)
+
 	fmt.Fprintln(w, "digraph DAG {\nrankdir=\"LR\"")
 
-	nodes := map[interface{}]bool{}
-	edges := map[pair]bool{}
+	printNodes(nodes, w)
+	printEdges(edges, w)
+
+	fmt.Fprintln(w, "}")
+}
+
+func collectDag(b dag.BuilderType) (nodes nodeCollection, edges edgeCollection) {
```

Comment:
we generally don't use named return parameters, let's drop the names here and declare them below.